### PR TITLE
CHAOSPLT-124: fix - update disruption target status with multiple kinds

### PR DESCRIPTION
**Pull request by aymericDD**
## What does this PR do?

- [ ] Adds new functionality
- [ ] Alters existing functionality
- [x] Fixes a bug
- [ ] Improves documentation or testing

Please briefly describe your changes as well as the motivation behind them:
- The target disruption status was not taking into account when a disruption injects multiple kinds of injections. Now the target status of the disruption lists the status of all its chaos pods.
- It also resolves another issue: in the case of a disruption with multiple types of injections, the disruption's status was marked as `injected` if only one of the two injections succeeded. The correct behavior is to mark the disruption status as `injected` only when all injections are successfully completed.

Example with the following disruption:

``` yaml
# Unless explicitly stated otherwise all files in this repository are licensed
# under the Apache License Version 2.0.
# This product includes software developed at Datadog (https://www.datadoghq.com/).
# Copyright 2023 Datadog, Inc.

apiVersion: chaos.datadoghq.com/v1beta1
kind: Disruption
metadata:
  name: network-drop
  namespace: chaos-demo
  annotations:
    chaos.datadoghq.com/environment: "lima"
spec:
  level: pod
  selector:
    app: demo-curl
  count: 100%
  network:
    drop: 100 # percentage of outgoing packets to drop
  diskFailure:
    paths:
      - /mnt/data/disk-read
      - /mnt/data/disk-write
    openat:
      exitCode: ENOENT
    probability: 100%
```

This disruption creates two chaos pods per target. One for the network disruption and one for the disk failure:

``` bash
$ kubectl get pods
NAME                               READY   STATUS    RESTARTS   AGE
chaos-controller-6b586f7b9-hg29g   2/2     Running   0          83s
chaos-network-drop-kv4zm           1/1     Running   0          60s
chaos-network-drop-zkcp8           1/1     Running   0          59s
```

> Before

_Status of the disruption_

``` bash
kubectl -n chaos-demo get disruption network-drop -ojson | jq .status                                                                                                                                                                                                   
```

``` json
{
  "desiredTargetsCount": 1,
  "ignoredTargetsCount": 0,
  "injectedTargetsCount": 1,
  "injectionStatus": "Injected",
  "selectedTargetsCount": 1,
  "targetInjections": {
    "demo-curl-bbc858869-z67wx": {
      "injectionStatus": "Injected",
      "injectorPodName": "chaos-network-drop-zkcp8",
      "since": "2023-10-30T16:48:11Z"
    }
  }
}
```

> After

_Status of the disruption_

``` bash
kubectl -n chaos-demo get disruption network-drop -ojson | jq .status                                                                                                                                                                                                   
```

``` json
{
  "desiredTargetsCount": 1,
  "ignoredTargetsCount": 0,
  "injectedTargetsCount": 1,
  "injectionStatus": "Injected",
  "selectedTargetsCount": 1,
  "targetInjections": {
    "demo-curl-bbc858869-z67wx": {
      "network-disruption": {
        "injectionStatus": "Injected",
        "injectorPodName": "chaos-network-drop-5flc4",
        "since": "2023-10-30T16:44:59Z"
      },
      "disk-failure": {
        "injectionStatus": "Injected",
        "injectorPodName": "chaos-network-drop-g9qjc",
        "since": "2023-10-30T16:45:10Z"
      }
    }
  }
}
```

## Code Quality Checklist

- [x] The documentation is up to date.
- [x] My code is sufficiently commented and passes continuous integration checks.
- [x] I have signed my commit (see [Contributing Docs](../CONTRIBUTING.md)).

## Testing

- [ ] I leveraged continuous integration testing
    - [ ] by depending on existing `unit` tests or `end-to-end` tests.
    - [ ] by adding new `unit` tests or `end-to-end` tests.
- [x] I manually tested the following steps:
    - [x] locally.
    - [ ] as a canary deployment to a cluster.

### CODE REVIEW

**Comment by ptnapoleon** [comment]
well, it's a little unfortunate that kubectl describe doesn't split up the output by chaos pod in any obvious way, but at least it is ordered correctly:
```
Status:
  Desired Targets Count:   1
  Ignored Targets Count:   0
  Injected Targets Count:  1
  Injection Status:        Injected
  Selected Targets Count:  1
  Target Injections:
    demo-curl-bbc858869-7fhpv:
      Disruption Kind Name:  network-disruption
      Injection Status:      Injected
      Injector Pod Name:     chaos-network-drop-ldcrg
      Since:                 2023-10-30T18:00:27Z
      Disruption Kind Name:  disk-failure
      Injection Status:      Injected
      Injector Pod Name:     chaos-network-drop-4nbj7
      Since:                 2023-10-30T18:00:37Z
```

I don't think we can fix that easily? we'd need to use the kind names as map keys, instead of items within the struct list. e.g.:
```
Status:
  Desired Targets Count:   1
  Ignored Targets Count:   0
  Injected Targets Count:  1
  Injection Status:        Injected
  Selected Targets Count:  1
  Target Injections:
    demo-curl-bbc858869-7fhpv:
      network-disruption:
          Injection Status:      Injected
          Injector Pod Name:     chaos-network-drop-ldcrg
          Since:                 2023-10-30T18:00:27Z
     disk-failure:
          Injection Status:      Injected
          Injector Pod Name:     chaos-network-drop-4nbj7
          Since:                 2023-10-30T18:00:37Z
      ```

File: api/v1beta1/disruption_types.go [hunk 1]
```diff
@@ -198,16 +198,38 @@ func (dd DisruptionDuration) Duration() time.Duration {
...
 	InjectionStatus chaostypes.DisruptionTargetInjectionStatus `json:"injectionStatus,omitempty"`
 	// since when this status is in place
 	Since metav1.Time `json:"since,omitempty"`
 }
 
+type TargetInjectorList []TargetInjection
+
 // TargetInjections map of target injection
-type TargetInjections map[string]TargetInjection
+type TargetInjections map[string]TargetInjectorList
```
Comment (ptnapoleon) [inline api/v1beta1/disruption_types.go:213]: What do you think of making this `map[string]map[string]TargetInjection?` Or alternatively, keeping this but changing `type TargetInjectorList` into `type TargetInjectorMap map[DisruptionKind]TargetInjection` in order to change the kubectl describe output to be a little nicer and separate the fields from separate injections

Comment (aymericDD) [reply api/v1beta1/disruption_types.go:213]: I agree with you, it will be less complex to parse 👍

File: api/v1beta1/disruption_types.go [hunk 2]
```diff
@@ -206,8 +206,25 @@ type TargetInjection struct {
 	Since metav1.Time `json:"since,omitempty"`
 }
 
+type TargetInjectorMap map[chaostypes.DisruptionKindName]TargetInjection
+
 // TargetInjections map of target injection
```
Comment (ptnapoleon) [inline api/v1beta1/disruption_types.go:211]:
```suggestion
// TargetInjections is a map of target names to injectors
```

**Review by ptnapoleon — DISMISSED** [review]
Could we deploy this in staging for a bit? but lgtm

File: api/v1beta1/disruption_types.go [hunk 3]
```diff
@@ -206,8 +206,25 @@ type TargetInjection struct {
...
+
+// GetInjectionWithDisruptionKind retrieves a TargetInjection associated with a specific DisruptionKindName from the map.
+//
+// Parameters:
+//   - kindName: The DisruptionKindName to search for in the map.
+//
+// Returns:
+//   - *TargetInjection: A pointer to the TargetInjection with the matching DisruptionKindName, or nil if not found.
+func (m TargetInjectorMap) GetInjectionWithDisruptionKind(kindName chaostypes.DisruptionKindName) *TargetInjection {
+	if targetInjection, isExist := m[kindName]; isExist {
```
Comment (Devatoria) [inline api/v1beta1/disruption_types.go:222]:
```suggestion
	if targetInjection, exists := m[kindName]; exists {
```

**Review by Devatoria — CHANGES_REQUESTED** [review]

**Review by ptnapoleon — APPROVED** [review]

File: controllers/disruption_controller.go [hunk 4]
```diff
@@ -697,11 +705,16 @@ func (r *DisruptionReconciler) handleChaosPodTermination(ctx context.Context, in
 }
 
 func (r *DisruptionReconciler) updateTargetInjectionStatus(instance *chaosv1beta1.Disruption, chaosPod corev1.Pod, status chaostypes.DisruptionTargetInjectionStatus, since metav1.Time) {
-	targetInjection := instance.Status.TargetInjections[chaosPod.Labels[chaostypes.TargetLabel]]
+	targetInjections := instance.Status.TargetInjections[chaosPod.Labels[chaostypes.TargetLabel]]
+
+	disruptionKindName := chaostypes.DisruptionKindName(chaosPod.Labels[chaostypes.DisruptionKindLabel])
+
+	targetInjection := targetInjections.GetInjectionWithDisruptionKind(disruptionKindName)
```
Comment (aymericDD) [inline controllers/disruption_controller.go:712]: The `targetInjection`could be `nil`. I have to handle this cases when the `targetInjection` variable is `nil`

**Review by ptnapoleon — APPROVED** [review]

**Review by Devatoria — APPROVED** [review]

File: controllers/disruption_controller.go [hunk 5]
```diff
@@ -493,8 +495,14 @@ func (r *DisruptionReconciler) startInjection(ctx context.Context, instance *cha
 				continue
 			}
 
-			if chaosv1beta1.ShouldSkipNodeFailureInjection(disKind, instance, injection) {
-				r.log.Debugw("skipping over injection, seems to be a re-injected node failure", "targetName", targetName, "injectionStatus", injection)
+			injection := injections.GetInjectionWithDisruptionKind(disKind)
+
+			if injection == nil {
+				return fmt.Errorf("the injection status from the target injections with this %s kind of disruption does not exist", disKind)
```
Comment (ptnapoleon) [inline controllers/disruption_controller.go:501]: when would we hit this condition?

Comment (aymericDD) [reply controllers/disruption_controller.go:501]: IMO I will said not, but if it's the case I prefered to have an explicit error rather an invalid memory access error which is more hard to debug

**Review by ptnapoleon — APPROVED** [review]

**Pull request merged**